File: python/src/lightcone_sdk/domain/market/wire.py
"""Market wire types - raw API response shapes."""

from dataclasses import dataclass, field
from typing import Optional

from ...error import _require


@dataclass
class OutcomeWire:
    """Raw outcome from the API."""
    index: int = 0
    name: str = ""
    icon_url_low: str = ""
    icon_url_medium: str = ""
    icon_url_high: str = ""

    @staticmethod
    def from_dict(d: dict, fallback_index: int = 0) -> "OutcomeWire":
        return OutcomeWire(
            index=d.get("index", fallback_index),
            name=d.get("name", ""),
            icon_url_low=d.get("icon_url_low", ""),
            icon_url_medium=d.get("icon_url_medium", ""),
            icon_url_high=d.get("icon_url_high", ""),
        )


@dataclass
class ConditionalMintWire:
    """Raw conditional mint nested inside a deposit asset."""
    id: int = 0
    token_address: str = ""
    outcome_index: int = 0
    outcome: str = ""
    short_symbol: str = ""
    symbol: str = ""
    description: Optional[str] = None
    decimals: int = 6
    icon_url_low: str = ""
    icon_url_medium: str = ""
    icon_url_high: str = ""

    @staticmethod
    def from_dict(d: dict) -> "ConditionalMintWire":
        return ConditionalMintWire(
            id=d.get("id", 0),
            token_address=d.get("token_address", ""),
            outcome_index=d.get("outcome_index", 0),
            outcome=d.get("outcome", ""),
            short_symbol=d.get("short_symbol", ""),
            symbol=d.get("symbol", ""),
            description=d.get("description"),
            decimals=d.get("decimals") or 6,
            icon_url_low=d.get("icon_url_low", ""),
            icon_url_medium=d.get("icon_url_medium", ""),
            icon_url_high=d.get("icon_url_high", ""),
        )


@dataclass
class DepositAssetWire:
    """Raw deposit asset from the API."""
    id: int = 0
    market_pubkey: str = ""
    deposit_asset: str = ""
    num_outcomes: int = 0
    display_name: str = ""
    token_symbol: str = ""
    symbol: str = ""
    description: Optional[str] = None
    decimals: int = 6
    icon_url_low: str = ""
    icon_url_medium: str = ""
    icon_url_high: str = ""
    conditional_mints: list[ConditionalMintWire] = field(default_factory=list)

    @staticmethod
    def from_dict(d: dict) -> "DepositAssetWire":
        return DepositAssetWire(
            id=d.get("id", 0),
            market_pubkey=d.get("market_pubkey", ""),
            deposit_asset=d.get("deposit_asset", ""),
            num_outcomes=d.get("num_outcomes", 0),
            display_name=d.get("display_name", ""),
            token_symbol=d.get("token_symbol", ""),
            symbol=d.get("symbol", ""),
            description=d.get("description"),
            decimals=d.get("decimals") or 6,
            icon_url_low=d.get("icon_url_low", ""),
            icon_url_medium=d.get("icon_url_medium", ""),
            icon_url_high=d.get("icon_url_high", ""),
            conditional_mints=[
                ConditionalMintWire.from_dict(cm)
                for cm in d.get("conditional_mints", [])
            ],
        )


@dataclass
class OrderbookWire:
    """Raw orderbook nested inside a market response."""
    id: int = 0
    market_pubkey: str = ""
    orderbook_id: str = ""
    base_token: str = ""
    quote_token: str = ""
    outcome_index: int = 0
    tick_size: int = 0
    total_bids: int = 0
    total_asks: int = 0
    last_trade_price: Optional[str] = None
    last_trade_time: Optional[str] = None
    active: bool = True

    @staticmethod
    def from_dict(d: dict) -> "OrderbookWire":
        return OrderbookWire(
            id=d.get("id", 0),
            market_pubkey=d.get("market_pubkey", ""),
            orderbook_id=d.get("orderbook_id", ""),
            base_token=d.get("base_token", ""),
            quote_token=d.get("quote_token", ""),
            outcome_index=d.get("outcome_index", 0),
            tick_size=d.get("tick_size", 0),
            total_bids=d.get("total_bids", 0),
            total_asks=d.get("total_asks", 0),
            last_trade_price=d.get("last_trade_price"),
            last_trade_time=d.get("last_trade_time"),
            active=d.get("active", True),
        )


@dataclass
class SearchOrderbook:
    """Minimal orderbook info returned from search endpoints."""
    orderbook_id: str = ""
    outcome_name: str = ""
    outcome_index: int = 0
    deposit_base_asset: str = ""
    deposit_quote_asset: str = ""
    deposit_base_symbol: str = ""
    deposit_quote_symbol: str = ""
    base_icon_url_low: str = ""
    base_icon_url_medium: str = ""
    base_icon_url_high: str = ""
    quote_icon_url_low: str = ""
    quote_icon_url_medium: str = ""
    quote_icon_url_high: str = ""
    conditional_base_mint: str = ""
    conditional_quote_mint: str = ""
    outcome_icon_url_low: Optional[str] = None
    outcome_icon_url_medium: Optional[str] = None
    outcome_icon_url_high: Optional[str] = None
    conditional_base_symbol: Optional[str] = None
    conditional_quote_symbol: Optional[str] = None
    latest_mid_price: Optional[str] = None

    @staticmethod
    def from_dict(d: dict) -> "SearchOrderbook":
        return SearchOrderbook(
            orderbook_id=d.get("orderbook_id", ""),
            outcome_name=d.get("outcome_name", ""),
            outcome_index=d.get("outcome_index", 0),
            deposit_base_asset=d.get("deposit_base_asset", ""),
            deposit_quote_asset=d.get("deposit_quote_asset", ""),
            deposit_base_symbol=d.get("deposit_base_symbol", ""),
            deposit_quote_symbol=d.get("deposit_quote_symbol", ""),
            base_icon_url_low=d.get("base_icon_url_low", ""),
            base_icon_url_medium=d.get("base_icon_url_medium", ""),
            base_icon_url_high=d.get("base_icon_url_high", ""),
            quote_icon_url_low=d.get("quote_icon_url_low", ""),
            quote_icon_url_medium=d.get("quote_icon_url_medium", ""),
            quote_icon_url_high=d.get("quote_icon_url_high", ""),
            conditional_base_mint=d.get("conditional_base_mint", ""),
            conditional_quote_mint=d.get("conditional_quote_mint", ""),
            outcome_icon_url_low=d.get("outcome_icon_url_low"),
            outcome_icon_url_medium=d.get("outcome_icon_url_medium"),
            outcome_icon_url_high=d.get("outcome_icon_url_high"),
            conditional_base_symbol=d.get("conditional_base_symbol"),
            conditional_quote_symbol=d.get("conditional_quote_symbol"),
            latest_mid_price=d.get("latest_mid_price"),
        )


@dataclass
class MarketWire:
    """Raw market data from the API."""
    market_id: int = 0
    market_pubkey: str = ""
    market_name: str = ""
    slug: Optional[str] = None
    description: Optional[str] = None
    definition: Optional[str] = None
    banner_image_url_low: Optional[str] = None
    banner_image_url_medium: Optional[str] = None
    banner_image_url_high: Optional[str] = None
    icon_url_low: Optional[str] = None
    icon_url_medium: Optional[str] = None
    icon_url_high: Optional[str] = None
    category: Optional[str] = None
    tags: list[str] = field(default_factory=list)
    featured_rank: Optional[int] = None
    market_status: Optional[str] = None
    winning_outcome: Optional[int] = None
    has_winning_outcome: bool = False
    volume: Optional[str] = None
    created_at: Optional[str] = None
    activated_at: Optional[str] = None
    settled_at: Optional[str] = None
    outcomes: list[OutcomeWire] = field(default_factory=list)
    deposit_assets: list[DepositAssetWire] = field(default_factory=list)
    orderbooks: list[OrderbookWire] = field(default_factory=list)
    # Derived from wire
    oracle: Optional[str] = None
    question_id: Optional[str] = None
    condition_id: Optional[str] = None

    @staticmethod
    def from_dict(d: dict) -> "MarketWire":
        return MarketWire(
            market_id=d.get("market_id", 0),
            market_pubkey=_require(d, "market_pubkey", "MarketWire"),
            market_name=d.get("market_name", ""),
            slug=d.get("slug"),
            description=d.get("description"),
            definition=d.get("definition"),
            banner_image_url_low=d.get("banner_image_url_low"),
            banner_image_url_medium=d.get("banner_image_url_medium"),
            banner_image_url_high=d.get("banner_image_url_high"),
            icon_url_low=d.get("icon_url_low"),
            icon_url_medium=d.get("icon_url_medium"),
            icon_url_high=d.get("icon_url_high"),
            category=d.get("category"),
            tags=d.get("tags") or [],
            featured_rank=d.get("featured_rank"),
            market_status=d.get("market_status"),
            winning_outcome=d.get("winning_outcome"),
            has_winning_outcome=d.get("has_winning_outcome", False),
            volume=d.get("volume"),
            created_at=d.get("created_at"),
            activated_at=d.get("activated_at"),
            settled_at=d.get("settled_at"),
            outcomes=[
                OutcomeWire.from_dict(o, fallback_index=i)
                for i, o in enumerate(d.get("outcomes", []))
            ],
            deposit_assets=[
                DepositAssetWire.from_dict(da)
                for da in d.get("deposit_assets", [])
            ],
            orderbooks=[
                OrderbookWire.from_dict(ob)
                for ob in d.get("orderbooks", [])
            ],
            oracle=d.get("oracle"),
            question_id=d.get("question_id"),
            condition_id=d.get("condition_id"),
        )


@dataclass
class MarketResponse:
    """API response for market list."""
    markets: list[MarketWire] = field(default_factory=list)
    next_cursor: Optional[int] = None
    has_more: bool = False

    @staticmethod
    def from_dict(d: dict) -> "MarketResponse":
        return MarketResponse(
            markets=[MarketWire.from_dict(m) for m in d.get("markets", [])],
            next_cursor=d.get("next_cursor"),
            has_more=d.get("has_more", False),
        )


@dataclass
class MarketSearchResult:
    """Search result from market search endpoints."""
    slug: str = ""
    market_name: str = ""
    market_status: Optional[str] = None
    category: Optional[str] = None
    tags: list[str] = field(default_factory=list)
    featured_rank: int = 0
    description: Optional[str] = None
    icon_url_low: Optional[str] = None
    icon_url_medium: Optional[str] = None
    icon_url_high: Optional[str] = None
    orderbooks: list[SearchOrderbook] = field(default_factory=list)

    @staticmethod
    def from_dict(d: dict) -> "MarketSearchResult":
        return MarketSearchResult(
            slug=d.get("slug", ""),
            market_name=d.get("market_name", ""),
            market_status=d.get("market_status"),
            category=d.get("category"),
            tags=d.get("tags") or [],
            featured_rank=d.get("featured_rank", 0),
            description=d.get("description"),
            icon_url_low=d.get("icon_url_low"),
            icon_url_medium=d.get("icon_url_medium"),
            icon_url_high=d.get("icon_url_high"),
            orderbooks=[
                SearchOrderbook.from_dict(ob)
                for ob in d.get("orderbooks", [])
            ],
        )


@dataclass
class MarketEvent:
    """WebSocket market event."""
    event_type: str = ""
    market_pubkey: str = ""
    status: Optional[str] = None
    winning_outcome: Optional[int] = None
    orderbook_id: Optional[str] = None

    @staticmethod
    def from_dict(d: dict) -> "MarketEvent":
        return MarketEvent(
            event_type=d.get("event_type", ""),
            market_pubkey=d.get("market_pubkey", ""),
            status=d.get("status"),
            winning_outcome=d.get("winning_outcome"),
            orderbook_id=d.get("orderbook_id"),
        )


@dataclass
class GlobalDepositAssetWire:
    """Raw globally whitelisted deposit asset from the API."""
    id: int = 0
    mint: str = ""
    display_name: Optional[str] = None
    symbol: Optional[str] = None
    description: Optional[str] = None
    icon_url_low: Optional[str] = None
    icon_url_medium: Optional[str] = None
    icon_url_high: Optional[str] = None
    decimals: Optional[int] = None
    whitelist_index: int = 0
    active: bool = False

    @staticmethod
    def from_dict(d: dict) -> "GlobalDepositAssetWire":
        return GlobalDepositAssetWire(
            id=d.get("id", 0),
            mint=_require(d, "mint", "GlobalDepositAssetWire"),
            display_name=d.get("display_name"),
            symbol=d.get("symbol"),
            description=d.get("description"),
            icon_url_low=d.get("icon_url_low"),
            icon_url_medium=d.get("icon_url_medium"),
            icon_url_high=d.get("icon_url_high"),
            decimals=d.get("decimals"),
            whitelist_index=d.get("whitelist_index", 0),
            active=d.get("active", False),
        )


@dataclass
class DepositMintsResponse:
    """API response for GET /api/markets/{market_pubkey}/deposit-assets."""

    market_pubkey: str = ""
    deposit_assets: list[DepositAssetWire] = field(default_factory=list)
    total: int = 0

    @staticmethod
    def from_dict(d: dict) -> "DepositMintsResponse":
        return DepositMintsResponse(
            market_pubkey=d.get("market_pubkey", ""),
            deposit_assets=[
                DepositAssetWire.from_dict(a)
                for a in d.get("deposit_assets", [])
            ],
            total=d.get("total", 0),
        )


@dataclass
class GlobalDepositAssetsListWire:
    """API response envelope for the global deposit asset whitelist."""
    assets: list[GlobalDepositAssetWire] = field(default_factory=list)
    total: int = 0

    @staticmethod
    def from_dict(d: dict) -> "GlobalDepositAssetsListWire":
        return GlobalDepositAssetsListWire(
            assets=[
                GlobalDepositAssetWire.from_dict(a)
                for a in d.get("assets", [])
            ],
            total=d.get("total", 0),
        )

File: rust/src/domain/market/wire.rs
//! Wire types for market responses (REST).

use std::collections::BTreeMap;

use crate::domain::market::Status;
use crate::domain::orderbook::wire::OrderbookResponse;
use crate::shared::{OrderBookId, PubkeyStr};
use chrono::{DateTime, Utc};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};

// ─── Outcome wire types ─────────────────────────────────────────────────────

/// Raw outcome from the REST API.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct OutcomeResponse {
    pub index: i16,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_url_low: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_url_medium: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_url_high: Option<String>,
}

// ─── Token wire types (REST) ────────────────────────────────────────────────

/// REST response for a deposit asset with its conditional mints.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct DepositAssetResponse {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub display_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token_symbol: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub symbol: Option<String>,
    pub deposit_asset: String,
    pub id: i32,
    pub market_pubkey: String,
    pub vault: String,
    pub num_outcomes: i16,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_url_low: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_url_medium: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_url_high: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata_uri: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub decimals: Option<i16>,
    pub conditional_mints: Vec<ConditionalTokenResponse>,
    pub created_at: DateTime<Utc>,
}

/// REST response for a conditional mint.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ConditionalTokenResponse {
    pub id: i32,
    pub outcome_index: i16,
    pub token_address: String,
    pub symbol: Option<String>,
    pub uri: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub outcome: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub deposit_symbol: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub short_symbol: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_url_low: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_url_medium: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_url_high: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata_uri: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub decimals: Option<i16>,
    pub created_at: DateTime<Utc>,
}

/// REST response for deposit mints list.
#[derive(Debug, Serialize, Deserialize, PartialEq)]
pub struct DepositMintsResponse {
    pub market_pubkey: String,
    pub deposit_assets: Vec<DepositAssetResponse>,
    pub total: usize,
}

// ─── Market resolution wire types ───────────────────────────────────────────

/// Canonical market resolution kind returned by the REST API.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum MarketResolutionKind {
    SingleWinner,
    Scalar,
}

/// Payout numerator for a single outcome in a resolved market.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct MarketResolutionPayout {
    pub outcome_index: i16,
    pub payout_numerator: i64,
}

/// Canonical payout-vector resolution returned by the REST API.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct MarketResolutionResponse {
    pub kind: MarketResolutionKind,
    pub payout_denominator: i64,
    pub payouts: Vec<MarketResolutionPayout>,
    pub single_winning_outcome: Option<i16>,
}

/// REST response for a single market.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct MarketResponse {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub market_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub slug: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub definition: Option<String>,
    pub outcomes: Vec<OutcomeResponse>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub banner_image_url_low: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub banner_image_url_medium: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub banner_image_url_high: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_url_low: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_url_medium: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_url_high: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub category: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub featured_rank: Option<i16>,
    pub market_pubkey: String,
    pub market_id: i64,
    pub oracle: String,
    pub question_id: String,
    pub condition_id: String,
    pub market_status: String,
    #[serde(default)]
    pub resolution: Option<MarketResolutionResponse>,
    pub created_at: DateTime<Utc>,
    pub activated_at: Option<DateTime<Utc>>,
    pub settled_at: Option<DateTime<Utc>>,
    pub deposit_assets: Vec<DepositAssetResponse>,
    pub orderbooks: Vec<OrderbookResponse>,
}

impl MarketResponse {
    pub fn is_resolved(&self) -> bool {
        self.resolution.is_some()
    }

    pub fn single_winning_outcome(&self) -> Option<i16> {
        self.resolution
            .as_ref()
            .and_then(|resolution| resolution.single_winning_outcome)
    }

    pub fn has_single_winning_outcome(&self) -> bool {
        self.single_winning_outcome().is_some()
    }
}

/// REST response for paginated markets list.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MarketsResponse {
    pub markets: Vec<MarketResponse>,
    pub next_cursor: Option<i64>,
    pub has_more: bool,
}

/// REST response wrapping a single market (used by by-slug and by-pubkey endpoints).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub(crate) struct SingleMarketResponse {
    pub market: MarketResponse,
}

/// Minimal search/featured result for a single orderbook.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SearchOrderbook {
    pub orderbook_id: OrderBookId,
    pub outcome_name: String,
    pub outcome_index: i16,
    pub deposit_base_asset: PubkeyStr,
    pub deposit_quote_asset: PubkeyStr,
    pub deposit_base_symbol: String,
    pub deposit_quote_symbol: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub base_icon_url_low: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub base_icon_url_medium: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub base_icon_url_high: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub quote_icon_url_low: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub quote_icon_url_medium: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub quote_icon_url_high: Option<String>,
    pub conditional_base_mint: PubkeyStr,
    pub conditional_quote_mint: PubkeyStr,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub outcome_icon_url_low: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub outcome_icon_url_medium: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub outcome_icon_url_high: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub conditional_base_symbol: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub conditional_quote_symbol: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub latest_mid_price: Option<Decimal>,
}

/// Minimal market result for search and featured listings.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct MarketSearchResult {
    pub slug: String,
    pub market_name: String,
    pub market_status: Status,
    pub category: Option<String>,
    #[serde(default)]
    pub tags: Vec<String>,
    pub featured_rank: i16,
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_url_low: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_url_medium: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_url_high: Option<String>,
    pub orderbooks: Vec<SearchOrderbook>,
}

/// Orderbooks for a single outcome within a market search result.
#[derive(Debug, Clone, PartialEq)]
pub struct SearchOutcomeGroup {
    pub outcome_index: i16,
    pub outcome_name: String,
    pub outcome_icon_url_low: Option<String>,
    pub outcome_icon_url_medium: Option<String>,
    pub outcome_icon_url_high: Option<String>,
    pub orderbooks: Vec<SearchOrderbook>,
    pub market_name: String,
    pub market_slug: String,
    pub market_icon_url_low: Option<String>,
    pub market_icon_url_medium: Option<String>,
    pub market_icon_url_high: Option<String>,
}

impl MarketSearchResult {
    pub fn orderbooks_by_outcome(&self) -> Vec<SearchOutcomeGroup> {
        let mut groups: BTreeMap<i16, SearchOutcomeGroup> = BTreeMap::new();
        for orderbook in &self.orderbooks {
            groups
                .entry(orderbook.outcome_index)
                .or_insert_with(|| SearchOutcomeGroup {
                    market_name: self.market_name.clone(),
                    market_slug: self.slug.clone(),
                    market_icon_url_low: self.icon_url_low.clone(),
                    market_icon_url_medium: self.icon_url_medium.clone(),
                    market_icon_url_high: self.icon_url_high.clone(),
                    outcome_index: orderbook.outcome_index,
                    outcome_name: orderbook.outcome_name.clone(),
                    outcome_icon_url_low: orderbook.outcome_icon_url_low.clone(),
                    outcome_icon_url_medium: orderbook.outcome_icon_url_medium.clone(),
                    outcome_icon_url_high: orderbook.outcome_icon_url_high.clone(),
                    orderbooks: Vec::new(),
                })
                .orderbooks
                .push(orderbook.clone());
        }
        groups.into_values().collect()
    }
}

// ─── Global deposit asset wire types ────────────────────────────────────────

/// REST response for a single globally whitelisted deposit asset.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct GlobalDepositAssetResponse {
    pub id: i32,
    pub mint: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub display_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub symbol: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_url_low: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_url_medium: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_url_high: Option<String>,
    pub decimals: Option<i16>,
    pub whitelist_index: i16,
    pub active: bool,
}

/// REST response envelope for the global deposit asset whitelist.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GlobalDepositAssetsListResponse {
    pub assets: Vec<GlobalDepositAssetResponse>,
    pub total: usize,
}

/// WS market lifecycle event.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "event_type")]
pub enum MarketEvent {
    #[serde(rename = "settled")]
    Settled { market_pubkey: String },
    #[serde(rename = "created")]
    Created { market_pubkey: String },
    #[serde(rename = "opened")]
    Opened { market_pubkey: String },
    #[serde(rename = "paused")]
    Paused { market_pubkey: String },
    #[serde(rename = "orderbook_created")]
    OrderbookCreated {
        market_pubkey: String,
        orderbook_id: String,
    },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn market_resolution_deserializes_single_winner() {
        let resolution: MarketResolutionResponse = serde_json::from_str(
            r#"{
                "kind": "single_winner",
                "payout_denominator": 1,
                "payouts": [
                    { "outcome_index": 0, "payout_numerator": 0 },
                    { "outcome_index": 1, "payout_numerator": 1 }
                ],
                "single_winning_outcome": 1
            }"#,
        )
        .unwrap();

        assert_eq!(resolution.kind, MarketResolutionKind::SingleWinner);
        assert_eq!(resolution.payout_denominator, 1);
        assert_eq!(resolution.single_winning_outcome, Some(1));
        assert_eq!(resolution.payouts[1].payout_numerator, 1);
    }

    #[test]
    fn market_resolution_deserializes_scalar() {
        let resolution: MarketResolutionResponse = serde_json::from_str(
            r#"{
                "kind": "scalar",
                "payout_denominator": 10,
                "payouts": [
                    { "outcome_index": 0, "payout_numerator": 7 },
                    { "outcome_index": 1, "payout_numerator": 3 }
                ],
                "single_winning_outcome": null
            }"#,
        )
        .unwrap();

        assert_eq!(resolution.kind, MarketResolutionKind::Scalar);
        assert_eq!(resolution.payout_denominator, 10);
        assert_eq!(resolution.single_winning_outcome, None);
        assert_eq!(resolution.payouts[0].payout_numerator, 7);
        assert_eq!(resolution.payouts[1].payout_numerator, 3);
    }
}
